### PR TITLE
Fix pynvml handles

### PR DESCRIPTION
**Pull request by quasiben**
Replaces https://github.com/dask/distributed/pull/8419

cc @fjetter @hendrikmakait @rjzamora 

I've tested this on machine with several A100s.  @jacobtomlinson is it possible for you to test this PR in a MIG setup ?

### CODE REVIEW

**Comment by jacobtomlinson** [comment]
I can, but with PyData London tomorrow I'm not going to get to this immediately.

**Comment by hendrikmakait** [comment]
> I've tested this on machine with several A100s. @jacobtomlinson is it possible for you to test this PR in a MIG setup ?

@jacobtomlinson: Gentle ping :)

**Comment by jacobtomlinson** [comment]
Apologies I've been struggling to get my hands on the right hardware to test this out. Let me ping NVIDIA folks again.

**Comment by hendrikmakait** [comment]
> Apologies I've been struggling to get my hands on the right hardware to test this out. Let me ping NVIDIA folks again.

No worries, I just wanted to make sure this didn't get buried in the post-conference catch-up.

**Comment by jacobtomlinson** [comment]
I'm seeing an exception when viewing the dashboard on a MIG machine with these changes.

Steps:

```bash
conda create -n test-mig -c conda-forge python=3.11 ipython -y
conda activate test-mig
pip install dask[complete]
pip install git+https://github.com/rapidsai/rapids-dask-dependency.git
pip install git+https://github.com/rapidsai/dask-cuda.git
pip install --no-deps git+https://github.com/dask/distributed.git@refs/pull/8693/head
ipython
```

```python
from dask_cuda import LocalCUDACluster
cluster = LocalCUDACluster()
```

Open dashboard and navigate to http://localhost:8787/gpu.

```pytb
2024-06-27 13:46:17,592 - distributed.dashboard.components.nvml - ERROR - int() argument must be a string, a bytes-like object or a real number, not 'NoneType'
Traceback (most recent call last):
  File "/raid/jtomlinson/miniforge3/envs/test-mig/lib/python3.11/site-packages/distributed/utils.py", line 838, in wrapper
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/raid/jtomlinson/miniforge3/envs/test-mig/lib/python3.11/site-packages/distributed/dashboard/components/nvml.py", line 135, in update
    utilization.append(int(u))
                       ^^^^^^
TypeError: int() argument must be a string, a bytes-like object or a real number, not 'NoneType'
2024-06-27 13:46:17,593 - distributed.dashboard.components.nvml - ERROR - int() argument must be a string, a bytes-like object or a real number, not 'NoneType'
Traceback (most recent call last):
  File "/raid/jtomlinson/miniforge3/envs/test-mig/lib/python3.11/site-packages/distributed/utils.py", line 838, in wrapper
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/raid/jtomlinson/miniforge3/envs/test-mig/lib/python3.11/site-packages/distributed/dashboard/components/nvml.py", line 185, in gpu_doc
    gpu_load.update()
  File "/raid/jtomlinson/miniforge3/envs/test-mig/lib/python3.11/site-packages/bokeh/core/property/validation.py", line 95, in func
    return input_function(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/raid/jtomlinson/miniforge3/envs/test-mig/lib/python3.11/site-packages/distributed/utils.py", line 838, in wrapper
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/raid/jtomlinson/miniforge3/envs/test-mig/lib/python3.11/site-packages/distributed/dashboard/components/nvml.py", line 135, in update
    utilization.append(int(u))
                       ^^^^^^
TypeError: int() argument must be a string, a bytes-like object or a real number, not 'NoneType'
2024-06-27 13:46:17,593 - tornado.application - ERROR - Uncaught exception GET /gpu (127.0.0.1)
HTTPServerRequest(protocol='http', host='localhost:8787', method='GET', uri='/gpu', version='HTTP/1.1', remote_ip='127.0.0.1')
Traceback (most recent call last):
  File "/raid/jtomlinson/miniforge3/envs/test-mig/lib/python3.11/site-packages/tornado/web.py", line 1790, in _execute
    result = await result
             ^^^^^^^^^^^^
  File "/raid/jtomlinson/miniforge3/envs/test-mig/lib/python3.11/site-packages/bokeh/server/views/doc_handler.py", line 54, in get
    session = await self.get_session()
              ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/raid/jtomlinson/miniforge3/envs/test-mig/lib/python3.11/site-packages/bokeh/server/views/session_handler.py", line 145, in get_session
    session = await self.application_context.create_session_if_needed(session_id, self.request, token)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/raid/jtomlinson/miniforge3/envs/test-mig/lib/python3.11/site-packages/bokeh/server/contexts.py", line 240, in create_session_if_needed
    self._application.initialize_document(doc)
  File "/raid/jtomlinson/miniforge3/envs/test-mig/lib/python3.11/site-packages/bokeh/application/application.py", line 190, in initialize_document
    h.modify_document(doc)
  File "/raid/jtomlinson/miniforge3/envs/test-mig/lib/python3.11/site-packages/bokeh/application/handlers/function.py", line 140, in modify_document
    self._func(doc)
  File "/raid/jtomlinson/miniforge3/envs/test-mig/lib/python3.11/site-packages/distributed/utils.py", line 838, in wrapper
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/raid/jtomlinson/miniforge3/envs/test-mig/lib/python3.11/site-packages/distributed/dashboard/components/nvml.py", line 185, in gpu_doc
    gpu_load.update()
  File "/raid/jtomlinson/miniforge3/envs/test-mig/lib/python3.11/site-packages/bokeh/core/property/validation.py", line 95, in func
    return input_function(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/raid/jtomlinson/miniforge3/envs/test-mig/lib/python3.11/site-packages/distributed/utils.py", line 838, in wrapper
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/raid/jtomlinson/miniforge3/envs/test-mig/lib/python3.11/site-packages/distributed/dashboard/components/nvml.py", line 135, in update
    utilization.append(int(u))
                       ^^^^^^
TypeError: int() argument must be a string, a bytes-like object or a real number, not 'NoneType'
```

File: distributed/diagnostics/nvml.py [hunk 1]
```diff
@@ -281,18 +316,22 @@ def get_device_mig_mode(device):
...
+def _get_handle(device):
+    try:
+        device_index = int(device)
+        return pynvml.nvmlDeviceGetHandleByIndex(device_index)
+    except ValueError:
+        uuid = device if isinstance(device, bytes) else bytes(device, "utf-8")
+        return pynvml.nvmlDeviceGetHandleByUUID(uuid)
+
+
 def _get_utilization(h):
```
Comment (jacobtomlinson) [inline distributed/diagnostics/nvml.py:335]: It looks like this method is returning `None` when using MIG devices which eventually results in the `int()` cast failing later on.

Comment (jacobtomlinson) [reply distributed/diagnostics/nvml.py:335]: It looks like this is to be expected. The documentation mentions that utilization rates is not supported on MIG https://docs.nvidia.com/deploy/nvml-api/group__nvmlDeviceQueries.html#group__nvmlDeviceQueries_1g540824faa6cef45500e0d1dc2f50b321

**Review by jacobtomlinson — APPROVED** [review]
I just pushed a fix for the exception I was seeing. I suspect the issue I was seeing has been here for a while regardless of the changes in this PR. 

Now that I can test these changes things seem to work as expected. Restricting `CUDA_VISIBLE_DEVICES` to one MIG partition by UUID works to I'm happy so merge this PR.

It seems that getting utilization isn't supported on MIG devices, so my fix that means those graphs will always show zero utilization. This may be surprising to users, so I'll raise a separate issue to track this.

**Comment by jacobtomlinson** [comment]
The GPU CI failure appears to be unrelated.

**Comment by jacobtomlinson** [comment]
Other failures seem unrelated. Merging.

**Pull request merged**